### PR TITLE
Improve relevancy of Quick Open Scene results

### LINKED ISSUE #50672: Quick open scene results are less relevant
### Godot version

e248d26

### System information

W10

### Issue description

I use Quick Open Scene quite a lot and in 4.0 I noticed that the results are less relevant (it might be related to numbers in the searched name).

I have a scene called DeadForest2, I search for it using "Dead2". In 3.x I get an accurate result
T2F68ZMJIq
but in 4.0 the scene that appears first is irrelevant to my search:
ARESSF1ere

### Steps to reproduce

1. Have many scenes
2. Use quick open to search for a scene
3. Compare results from 3.x and 4.0

### Minimal reproduction project

[ReproductionProject.zip](https://github.com/godotengine/godot/files/6851725/ReproductionProject.zip)
The project contains that particular bugged case from the GIF.

### PR DESCRIPTION
**Pull request by SirQuartz**
This pull request improves the accuracy of the quick open scene results.

Resolves: #50672

https://user-images.githubusercontent.com/62965063/126517931-b112b265-883b-4d84-a6b5-33768fb0a5cc.mp4

### CODE REVIEW

File: editor/quick_open.cpp [hunk 1]
```diff
@@ -118,6 +118,11 @@ void EditorQuickOpen::_update_search() {
 float EditorQuickOpen::_score_path(const String &p_search, const String &p_path) {
 	float score = 0.9f + .1f * (p_search.length() / (float)p_path.length());
 
+	// Exact match.
+	if (p_search == p_path) {
+		return 1.2f;
```
Comment (Calinou) [inline editor/quick_open.cpp:123]: Is that `1.2` a magic number somehow? There should be a comment to describe why that particular value is returned.

Comment (SirQuartz) [reply editor/quick_open.cpp:123]: I have no idea why it's special, I just compared the quick open scene of 3.x and 4.0 and found that it was missing this if statement. All I knew was that things worked as they should in 3.x, so this seems to have replicated that.

**Comment by akien-mga** [comment]
I swear this code is being changed at every other release... Would be good to figure out why it regressed, especially if the fix is to readd the same code as in `3.x`.

**Comment by SirQuartz** [comment]
> I swear this code is being changed at every other release... Would be good to figure out why it regressed, especially if the fix is to readd the same code as in `3.x`.

Yeah, I really don't know why it would have regressed in the first place. As far as I could tell it worked perfectly fine in 3.x. It also seems to work mostly fine with a similar implementation in 4.0 as I've done here. 

As far as I can tell, the changes I've made improve its relevancy and there appear to be no noticeable problems.

Here's a slightly more complex project layout with multiple folders and scenes spread across them randomly:

https://user-images.githubusercontent.com/62965063/126715301-41f28a1b-f5d5-47cd-898f-720b89091a51.mp4

All around it handles it very well, it's accurate and even works with spelling mistakes, so this seems fine.

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot/issues/46427.

**Comment by akien-mga** [comment]
What I meant is that if the code from `3.x` works better, it means someone changed in `master` to make it regress. So we can ask the `git log` to know who:

https://github.com/godotengine/godot/commit/c33707e9b44478f6efaf57a0ca9c8496bcb87695#diff-ee274d6f77452fe1dbc2fe0d16b89a7599516564ded336f2060bdd21dea8343a
https://github.com/godotengine/godot/commit/7f6e8d25d43c77b89bd3a9968e13ebef94a82f5c#diff-ee274d6f77452fe1dbc2fe0d16b89a7599516564ded336f2060bdd21dea8343a
https://github.com/godotengine/godot/commit/269ed93271bc3e6a6eaecc6db4e1bfbcec69c1a8#diff-ee274d6f77452fe1dbc2fe0d16b89a7599516564ded336f2060bdd21dea8343a

CC @hinlopen

**Comment by hinlopen** [comment]
Seems like it might be because it finds a subsequence that starts earlier in a folder (Worl**d**5/Tw**e**enl**a**n**d**2).

Anyway, the search ranking was changed because of #39927. Just changing it back will likely reintroduce that.

**Comment by SirQuartz** [comment]
> Seems like it might be because it finds a subsequence that starts earlier in a folder (Worl**d**5/Tw**e**enl**a**n**d**2).
> 
> Anyway, the search ranking was changed because of #39927. Just changing it back will likely reintroduce that.

@akien-mga @hinlopen I'm pretty sure a better implementation of this that would incorporate the original implementation in 3.x and the fix done by #39927 would be to just have it ignore file extensions altogether for the path when comparing the similarities.

So something like:

``` C++
// Similarity
	return p_path.to_lower().get_basename().similarity(p_search.to_lower());
```
Having the `.get_basename()` ensures that it doesn't include the extension as part of the check for similarity.

Anyway here's me demonstrating this change:


https://user-images.githubusercontent.com/62965063/126779924-33556a8c-8677-4f99-bfae-7fc897cca506.mp4


https://user-images.githubusercontent.com/62965063/126779946-4192774e-8631-4ed7-b1a9-bb856736de06.mp4

In all 3 of the different quick open search types that I've tested it works perfectly fine with this implementation, and again, is very accurate with its search results. It now favors file names over file extensions in a blind search, which is more useful in almost every situation. But you can still explicitly search for file extensions if you want and it will show them too. So I've added that change to this commit now.

**Comment by KoBeWi** [comment]
btw 3.x implementation isn't perfect either, #51210 doesn't exist on master. This PR seems to fix all of the issues so far.

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**